pkg/bundle/fs.go

```diff
@@ -0,0 +1,199 @@
+	scheme := runtime.NewScheme()
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(operatorsv1alpha1.AddToScheme(scheme))
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	utilruntime.Must(rukpakv1alpha1.AddToScheme(scheme))
+
+	fsys := FS{
+		baseFS:     baseFS,
+		scheme:     scheme,
+		storedObjs: make(map[string]ObjectFile[client.Object]),
+	}
+	for _, opt := range opts {
+		opt(&fsys)
+	}
+
+	return fsys
+}
+
+// WithManifestDir returns an option for creating new bundle filesystems.
+//
+// This option specifies which directories can contain manifests. If this opion is unset then the filesystem will search
+// in all directories for manifests.
+func WithManifestDirs(dirs ...string) func(*FS) {
+	return func(fsys *FS) {
+		fsys.manifestDirs = append(fsys.manifestDirs, dirs...)
+	}
+}
+
+// WithExtraTypes returns an option for creating new bundle filesystems.
+//
+// This option takes in functions that mutate the scheme. The scheme by default only contains the Kubernetes builtin
+// types and the types required by known bundle formats. The functions provided here will be applied to the scheme to
+// add additional types. The expected usage is to add the `AddToScheme` functions from generated type packages.
+func WithExtraTypes(addToSchemes []func(s *runtime.Scheme) error) func(*FS) {
+	return func(fsys *FS) {
+		for _, addToScheme := range addToSchemes {
+			utilruntime.Must(addToScheme(fsys.scheme))
+		}
+	}
+}
+
+// WithStrictTypes returns an option for creating new bundle filesystems.
+//
+// This option enables strict type checking for the created bundle filesystem. When strict type checking is enabled,
+// the filesystem will throw an error when a caller attempts to open a file that has a manifest of a type unknown to the
+// scheme. When strict type checking is disabled, those unknown manifests will be left unstructured.
+func WithStrictTypes() func(*FS) {
+	return func(fsys *FS) {
+		fsys.strictTypes = true
+	}
+}
+
+// Open a file if it exists on the file system.
+//
+// If the file is in a directory that contains manifests and ends in `.yaml` or `.yml` then it will return an
+// `*ObjectFile` with the objects parsed to their concrete go type.
+func (fsys FS) Open(name string) (fs.File, error) {
+	// TODO(ryantking): When we figure out how to store objects, this will change
+	if objFile, ok := fsys.storedObjs[name]; ok {
+		return objFile, nil
+	}
+
+	f, err := fsys.baseFS.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	if !fsys.isManifestFile(name) {
+		return f, nil
+	}
+
+	objFile, err := NewObjectFile[client.Object](f, fsys.scheme, fsys.strictTypes)
+	if err != nil {
+		f.Close() // close the file that's already open
+		return nil, err
+	}
+
+	return objFile, nil
+}
+
+func (fsys FS) isManifestFile(name string) bool {
+	if filepath.Ext(name) != ".yaml" || filepath.Ext(name) == ".yml" {
+		return false
+	}
+	if len(fsys.manifestDirs) == 0 {
+		return true
+	}
```

Comment:
In the case of a theoretical helm chart bundle, I assume there would be no manifest dirs, which would mean every file _is_ a manifest file? Should this block actually return false?